### PR TITLE
update to a-c 0.33.0

### LINKED ISSUE #189: Upgrade to Android-Components 0.32.0
This task will involve updating the use of the `fxa-client` library to reflect breaking API changes, as well as adding specs for the `AccountStore`.

## Todo

* [x] update gradle dependencies
* [x] make changes where breakage occurs
* [x] revisit `app/lint-baseline.xml`

Depends on mozilla-mobile/android-components#1332

### PR DESCRIPTION
**Pull request by sashei**
Fixes #189 

## To Do

- [x] add “WIP” to the PR title if pushing up but not complete nor ready for review
- [x] double check the original issue to confirm it is fully satisfied
- [x] add testing notes and screenshots in PR description to help guide reviewers
- [x] add unit tests
- [x] consider running this branch in a debug simulator and check for memory leak notifications or any warnings
- [x] make sure CI builds are passing (e.g.: fix lint and other errors)

### CODE REVIEW

**Comment by sashei** [comment]
This is blocked on the upgrade of the `service-sync-logins` component using the most recent version of Kotlin / coroutines. We're getting the following failure when trying to use the `sync-logins` library from our app:

```
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: mozilla.lockbox, PID: 18651
    java.lang.NoClassDefFoundError: Failed resolution of: Lkotlinx/coroutines/experimental/BuildersKt;
        at org.mozilla.sync15.logins.MemoryLoginsStorage.asyncResult(MemoryLoginsStorage.kt:218)
        at org.mozilla.sync15.logins.MemoryLoginsStorage.isLocked(MemoryLoginsStorage.kt:60)
        at mozilla.lockbox.store.DataStore.unlock(DataStore.kt:90)
        at mozilla.lockbox.store.DataStore$2.accept(DataStore.kt:69)
        at mozilla.lockbox.store.DataStore$2.accept(DataStore.kt:24)
        at io.reactivex.internal.observers.LambdaObserver.onNext(LambdaObserver.java:63)
        at io.reactivex.internal.operators.observable.ObservableMap$MapObserver.onNext(ObservableMap.java:62)
        at io.reactivex.internal.operators.observable.ObservableFilter$FilterObserver.onNext(ObservableFilter.java:52)
        at io.reactivex.internal.operators.observable.ObservableFlatMap$MergeObserver.tryEmit(ObservableFlatMap.java:265)
        at io.reactivex.internal.operators.observable.ObservableFlatMap$InnerObserver.onNext(ObservableFlatMap.java:572)
        at io.reactivex.subjects.PublishSubject$PublishDisposable.onNext(PublishSubject.java:308)
        at io.reactivex.subjects.PublishSubject.onNext(PublishSubject.java:228)
        at mozilla.lockbox.flux.Dispatcher.dispatch(Dispatcher.kt:22)
        at mozilla.lockbox.presenter.ItemListPresenter.onViewReady(ItemListPresenter.kt:89)
        at mozilla.lockbox.view.CommonFragment.onViewCreated(CommonFragment.kt:19)
        at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1471)
        at android.support.v4.app.FragmentManagerImpl.addAddedFragments(FragmentManager.java:2646)
        at android.support.v4.app.FragmentManagerImpl.executeOpsTogether(FragmentManager.java:2416)
        at android.support.v4.app.FragmentManagerImpl.removeRedundantOperationsAndExecute(FragmentManager.java:2372)
        at android.support.v4.app.FragmentManagerImpl.execPendingActions(FragmentManager.java:2273)
        at android.support.v4.app.FragmentManagerImpl$1.run(FragmentManager.java:733)
        at android.os.Handler.handleCallback(Handler.java:873)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6669)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
E/AndroidRuntime: Caused by: java.lang.ClassNotFoundException: Didn't find class "kotlinx.coroutines.experimental.BuildersKt" on path: DexPathList[[zip file "/system/framework/org.apache.http.legacy.boot.jar", zip file "/data/app/mozilla.lockbox-a05zebk55jBK5kCVjBX4tw==/base.apk", zip file "/data/app/mozilla.lockbox-a05zebk55jBK5kCVjBX4tw==/split_lib_dependencies_apk.apk", zip file "/data/app/mozilla.lockbox-a05zebk55jBK5kCVjBX4tw==/split_lib_resources_apk.apk", zip file "/data/app/mozilla.lockbox-a05zebk55jBK5kCVjBX4tw==/split_lib_slice_0_apk.apk", zip file "/data/app/mozilla.lockbox-a05zebk55jBK5kCVjBX4tw==/split_lib_slice_1_apk.apk", zip file "/data/app/mozilla.lockbox-a05zebk55jBK5kCVjBX4tw==/split_lib_slice_2_apk.apk", zip file "/data/app/mozilla.lockbox-a05zebk55jBK5kCVjBX4tw==/split_lib_slice_3_apk.apk", zip file "/data/app/mozilla.lockbox-a05zebk55jBK5kCVjBX4tw==/split_lib_slice_4_apk.apk", zip file "/data/app/mozilla.lockbox-a05zebk55jBK5kCVjBX4tw==/split_lib_slice_5_apk.apk", zip file "/data/app/mozilla.lockbox-a05zebk55jBK5kCVjBX4tw==/split_lib_slice_6_apk.apk", zip file "/data/app/mozilla.lockbox-a05zebk55jBK5kCVjBX4tw==/split_lib_slice_7_apk.apk", zip file "/data/app/mozilla.lockbox-a05zebk55jBK5kCVjBX4tw==/split_lib_slice_8_apk.apk", zip file "/data/app/mozilla.lockbox-a05zebk55jBK5kCVjBX4tw==/split_lib_slice_9_apk.apk"],nativeLibraryDirectories=[/data/app/mozilla.lockbox-a05zebk55jBK5kCVjBX4tw==/lib/x86, /data/app/mozilla.lockbox-a05zebk55jBK5kCVjBX4tw==/base.apk!/lib/x86, /data/app/mozilla.lockbox-a05zebk55jBK5kCVjBX4tw==/split_lib_dependencies_apk.apk!/lib/x86, /data/app/mozilla.lockbox-a05zebk55jBK5kCVjBX4tw==/split_lib_resources_apk.apk!/lib/x86, /data/app/mozilla.lockbox-a05zebk55jBK5kCVjBX4tw==/split_lib_slice_0_apk.apk!/lib/x86, /data/app/mozilla.lockbox-a05zebk55jBK5kCVjBX4tw==/split_lib_slice_1_apk.apk!/lib/x86, /data/app/mozilla.lockbox-a05zebk55jBK5kCVjBX4tw==/split_lib_slice_2_apk.apk!/lib/x86, /data/app/mozilla.lockbox-a05zebk55jBK5kCVjBX4tw==/split_lib_slice_3_apk.apk!/lib/x86, /data/app/mozilla.lockbox-a05zebk55jBK5kCVjBX4tw==/split_lib_slice_4_apk.apk!/lib/x86, /data/app/mozilla.lockbox-a05zebk55jBK5kCVjBX4tw==/split_lib_slice_5_apk.apk!/lib/x86, /data/app/mozilla.lockbox-a05zebk55jBK5kCVjBX4tw==/split_lib_slice_6_apk.apk!/lib/x86, /data/app/mozilla.lockbox-a05zebk55jBK5kCVjBX4tw==/split_lib_slice_7_apk.apk!/lib/x86, /data/app/mozilla.lockbox-a05zebk55jBK5kCVjBX4tw==/split_lib_slice_8_apk.apk!/lib/x86, /data/app/mozilla.lockbox-a05zebk55jBK5kCVjBX4tw==/split_lib_slice_9_apk.apk!/lib/x86, /system/lib]]
        at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:134)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
        	... 28 more
    	Suppressed: java.io.IOException: No original dex files found for dex location /data/app/mozilla.lockbox-a05zebk55jBK5kCVjBX4tw==/split_lib_resources_apk.apk
        at dalvik.system.DexFile.openDexFileNative(Native Method)
        at dalvik.system.DexFile.openDexFile(DexFile.java:354)
        at dalvik.system.DexFile.<init>(DexFile.java:101)
        at dalvik.system.DexFile.<init>(DexFile.java:75)
        at dalvik.system.DexPathList.loadDexFile(DexPathList.java:394)
        at dalvik.system.DexPathList.makeDexElements(DexPathList.java:354)
        at dalvik.system.DexPathList.<init>(DexPathList.java:164)
        at dalvik.system.BaseDexClassLoader.<init>(BaseDexClassLoader.java:74)
        at dalvik.system.BaseDexClassLoader.<init>(BaseDexClassLoader.java:65)
        at dalvik.system.PathClassLoader.<init>(PathClassLoader.java:64)
        at com.android.internal.os.ClassLoaderFactory.createClassLoader(ClassLoaderFactory.java:73)
        at com.android.internal.os.ClassLoaderFactory.createClassLoader(ClassLoaderFactory.java:88)
        at android.app.ApplicationLoaders.getClassLoader(ApplicationLoaders.java:74)
        at android.app.ApplicationLoaders.getClassLoader(ApplicationLoaders.java:40)
        at android.app.LoadedApk.createOrUpdateClassLoaderLocked(LoadedApk.java:727)
        at android.app.LoadedApk.getClassLoader(LoadedApk.java:810)
        at android.app.LoadedApk.getResources(LoadedApk.java:1032)
        at android.app.ContextImpl.createAppContext(ContextImpl.java:2345)
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:5749)
        at android.app.ActivityThread.access$1100(ActivityThread.java:199)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1650)
        at android.os.Handler.dispatchMessage(Handler.java:106)
```

**Comment by linuxwolf** [comment]
CC @thomcc
This seems like a mixup of dependencies; "us" (lockbox) vs "them" (android-components).  Any quick insights?

**Comment by thomcc** [comment]
Huh, you might need `implementation 'org.jetbrains.kotlinx:kotlinx-coroutines-android:0.23.4'` (or something similar) in your build.gradle.

Once https://github.com/mozilla-mobile/android-components/pull/1332 is available (should be very soon), this probably shouldn't be an issue (although it's a breaking change).

**Comment by linuxwolf** [comment]
> Huh, you might need `implementation 'org.jetbrains.kotlinx:kotlinx-coroutines-android:0.23.4'` (or something similar) in your build.gradle.

We added that, but version `1.0.1` instead of `0.23.4`. 🤔 

> 
> Once [mozilla-mobile/android-components#1332](https://github.com/mozilla-mobile/android-components/pull/1332) is available (should be very soon), this probably shouldn't be an issue (although it's a breaking change).

Understood about breaking API.  We'll put this PR on hold until that is available.

**Comment by thomcc** [comment]
> yes, and `service-fxa-client` requires `1.0.1`. as I cited above, the `service-sync-logins` needs to be on the same version for us to upgrade & use both from `android-components`.

This is a little surprising, since the previous version of logins worked in the sample inside A-C.

**Comment by sashei** [comment]
@thomcc I thought so too! looking at `Dependencies.kt` I would think that both would end up on `1.0.1`...

**Comment by sashei** [comment]
@thomcc maybe there's something in the library compilation / vendoring that I don't understand here

**Comment by thomcc** [comment]
Hmm, the 0.23.4 version is coming from https://github.com/mozilla/application-services/blob/v0.8.1/logins-api/android/library/build.gradle#L117 (This is the 0.8.1 release -- the most recent one moves the coroutine code to A-C).

So it makes sense that you'd have both 0.23.4 and 1.0.1, but I think that should still work -- if it didn't I think we'd have had issues with https://github.com/mozilla-mobile/android-components/blob/master/samples/sync-logins/build.gradle, which used both logins (and thus it's 0.23.4 coroutines lib) and fxa (and thus it's 1.0.1 coroutines lib)

**Comment by sashei** [comment]
@thomcc I'm trying out that sample app locally; the crash I got happened at runtime not compile time, so maybe there's a more subtle problem there? it's also possible that the sample app does not yet use Kotlin 1.3, as I think that's where the issue arises based on this post: https://discuss.kotlinlang.org/t/feedback-wanted-migration-to-stable-coroutines-in-kotlin-1-3/8849

basically, the `experimental` category from the coroutines library with a version < `1.0` will not be supported if your app is on Kotlin 1.3.

**Comment by sashei** [comment]
so maybe the solution here is to downgrade the kotlin version...

**Comment by thomcc** [comment]
Ah, yeah, that would make sense then. Then yeah, this should be fixed by https://github.com/mozilla-mobile/android-components/pull/1332 (which just landed), but you might be out of luck until that goes out. If you really want to make progress you can follow steps like https://mozilla-mobile.github.io/android-components/contributing/testing-components-inside-app to use a-c master locally.

**Comment by linuxwolf** [comment]
I admit to calling out the version diffs because of concerns if upstream uses different versions of the same library.  Note the sample-apps are compiled/assembled, but not _tested_ in CI.

We'll be waiting for mozilla-mobile/anroid-components#1332 to make into a release, and I think this whole thread highlights my concerns.

**Comment by thomcc** [comment]
> I admit to calling out the version diffs because of concerns if upstream uses different versions of the same library. Note the sample-apps are compiled/assembled, but not tested in CI.

Right, it was tested manually though.

Anyway, the version diff is one of the reasons why we did 1332 sooner rather than later. I think the only current dependency shared between the two is the kotlin stdlib, which probably does have a risk of being out of sync, but it's tricky to balance this between avoiding having unsafe code in android-components (we'd like to avoid this because it makes ABI-breaking changes on our end very risky, since nothing prevents them from accidentally using the wrong version and risking arbitrary problems).

I guess a-s could do things in Java (and not Kotlin) and avoid that dependency, and maybe that's the right choice, given all of this.

File: app/src/main/java/mozilla/lockbox/LockboxApplication.kt [hunk 1]
```diff
@@ -39,25 +38,16 @@ class LockboxApplication : Application() {
     override fun onCreate() {
         super.onCreate()
         if (leakCanary()) return
+        if (isUnitTest()) return
```
Comment (sashei) [inline app/src/main/java/mozilla/lockbox/LockboxApplication.kt]: for some reason, in the new Robolectric world this onCreate method gets called every time. I had to include this call to keep the real versions of the .shared Stores from getting allocated and accessing things outside of the scope of the test.

File: app/src/main/java/mozilla/lockbox/adapter/SettingCellConfiguration.kt [hunk 2]
```diff
@@ -15,11 +15,10 @@ abstract class SettingCellConfiguration(
     @StringRes open val subtitle: Int? = null
 )
 
-@Suppress("WRONG_ANNOTATION_TARGET_WITH_USE_SITE_TARGET_ON_TYPE")
 class TextSettingConfiguration(
     @StringRes override val title: Int,
     @StringRes override val subtitle: Int? = null,
-    val detailTextDriver: Observable<@param:StringRes Int>,
```
Comment (sashei) [inline app/src/main/java/mozilla/lockbox/adapter/SettingCellConfiguration.kt]: no longer works with Kotlin 1.3 :(

File: app/src/main/res/values/strings.xml [hunk 3]
```diff
@@ -71,7 +71,6 @@
         from Firefox Lockbox.
     </string>
     <string name="disconnect">Disconnect</string>
-    <string name="error">Fingerprint sensor error</string>
```
Comment (devinreams) [inline app/src/main/res/values/strings.xml]: Nice 🛀

File: gradle/wrapper/gradle-wrapper.properties [hunk 4]
```diff
@@ -1,12 +1,6 @@
-#
-# This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (devinreams) [inline gradle/wrapper/gradle-wrapper.properties]: Do you think it's worth manually adding back the copyright notice in this properties file?

Comment (linuxwolf) [reply gradle/wrapper/gradle-wrapper.properties]: I'm assuming this file was automatically generated/updated?  If so, there's little point in adding back the header as it'll disappear the next time it changes.

Comment (sashei) [reply gradle/wrapper/gradle-wrapper.properties]: ah, that's a good point... I don't remember manually adding this gradle change, so likely it's automatically generated :p

**Pull request merged**